core/src/main/scala/org/apache/spark/deploy/history/BasicEventFilterBuilder.scala

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.history
+
+import scala.collection.mutable
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.scheduler._
+
+class BasicEventFilterBuilder extends SparkListener with EventFilterBuilder {
```

Comment:
I think it's a weird construct. `BasicEventFilterBuilder` should build a filter with is `SparkListener`. Same applies to the other builder.

Reply:
So no luck, as SparkListener is not a trait, but an abstract class.

Reply:
I've invested time to understand your point. The following worked for me:
```
...
private[spark] class BasicEventFilterBuilder extends EventFilterBuilder
...
private[spark] trait EventFilter extends SparkListenerInterface
...
private[spark] class BasicEventFilter extends SparkListener with EventFilter with Logging
...
```
Testing would be a bit harder though...

Reply:
BasicEventFilterBuilder should extend SparkListener; it is optional for BasicEventFilter to extend SparkListener(Interface) but it is intended for BasicEventFilterBuilder to extend SparkListener. BasicEventFilterBuilder and BasicEventFilter are working for different purposes.

Reply:
> BasicEventFilterBuilder should extend SparkListener

Why? Not yet see your point. Is it not working what I've suggested?

If we consider responsibilities:
* `SparkListener` is input
* `EventFilterBuilder` builds `EventFilter`
* `EventFilter` filters input to output
Adding `EventFilterBuilder` input functionality is side effect.

Reply:
Asking it differently why `EventFilter` can't build up itself from `SparkListener` events and why can't be this list of `EventFilter` passed to `FilteredEventLogFileRewriter`?

Reply:
> why EventFilter can't build up itself from SparkListener events

It's possible but then the filter does too many things and UTs will be harder to implement. Please refer how I craft UTs for filters - it passes the information which is expected to be passed from filter builder. If we unify two classes, not intuitive to inject the test information.

Reply:
Back to the origin question, SparkListener interface defines "what to do" when the event comes in. EventFilter interface defines the "predicate" whether the event should be filtered or not when the event comes in.

```
...
private[spark] class BasicEventFilterBuilder extends EventFilterBuilder
...
private[spark] trait EventFilter extends SparkListenerInterface
...
private[spark] class BasicEventFilter extends SparkListener with EventFilter with Logging
...
```

So `private[spark] trait EventFilter extends SparkListenerInterface` is not same as current one and doesn't match with my intention. Hope this clarify my intention.

Reply:
> It's possible but then the filter does too many things

I can relate to this.

Now I see your point and I have the feeling both are right.
`Builder` is needed to create type specific constructs.
`EventFilter` is needed as predicate.
Maybe we can create `EventSource` (or whatever name which makes more sense).

```
private[spark] class BasicEventFilterBuilder extends EventFilterBuilder
private[spark] trait EventSource extends SparkListener
private[spark] trait EventFilter {
  val eventSource: EventSource
}
private[spark] class BasicEventSource extends EventSource
private[spark] class BasicEventFilter extends EventFilter with Logging
```